Logging fixes

trexfeathers opened the pull request:
## 🚀 Pull Request

### Description
<!-- Provide a clear description about your awesome pull request -->
<!-- Tell us all about your new feature, improvement, or bug fix -->

Improvements to all uses of logging in `experimental.ugrid`, following recent team discoveries.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

trexfeathers commented on lib/iris/experimental/ugrid/__init__.py:
```diff
@@ -3673,7 +3689,6 @@ def _build_mesh(cf, mesh_var, file_path):
             ", consistent with the attached connectivities."
         )
         # TODO: reconsider logging level when we have consistent practice.
-        # TODO: logger always requires extras['cls'] : can we fix this?
```
`extra` is useful for almost all examples in this module, excepting those in `_build_mesh()`. AFAICT the only way to not require this kwarg, while still offering the benefits when wanted, would be something like this at the top of the module:

```Python
logger._log = functools.partial(logger._log, extra=dict(cls=None))
```

... and that feels pretty hacky, right?

trexfeathers commented on lib/iris/experimental/ugrid/__init__.py:
```diff
@@ -3687,7 +3702,6 @@ def _build_mesh(cf, mesh_var, file_path):
                 " -- ignoring this as it is inconsistent."
             )
             # TODO: reconsider logging level when we have consistent practice.
-            # TODO: logger always requires extras['cls'] : can we fix this?
```
https://github.com/SciTools/iris/pull/4106/files#r618543039

pp-mo commented:
Useful improvements you've made here.
But on the testing side, I couldn't avoid thinking about how to improve this so I went digging and experimenting.
I came up with a possible way here : https://github.com/pp-mo/iris/pull/68/files

Don't want to undermine the work here, but I think the test changes are looking messy.
Would you consider incorporating into this PR the above extension to the `assertLogs` method ??
I do still have some concerns about the completeness of that aproach, though : see https://github.com/pp-mo/iris/pull/68

trexfeathers commented:
>Don't want to undermine the work here, but I think the test changes are looking messy.

3cf66627856db2214eeeae05ff4a61a349172449  Does this help at all?

>Would you consider incorporating into this PR the above extension to the assertLogs method ??

bfbed13737fe3099aef67864a26cd8dd38b30e96  I can confirm from local testing that this addresses the issue you originally identified.

>I do still have some concerns about the completeness of that aproach, though

I get what you're saying, however this area is new to us and we have addressed all issues we've encountered so far. No problem IMO.

pp-mo commented on lib/iris/tests/__init__.py:
```diff
@@ -607,6 +607,31 @@ def assertWarnsRegexp(self, expected_regexp=""):
         self.assertTrue(matches, msg)
 
+    @contextlib.contextmanager
+    def assertLogs(self, logger=None, level=None):
```
```suggestion
    def assertLogs(self, logger=None, level=None, msg_regex=None):
```

pp-mo commented on lib/iris/tests/__init__.py:
```diff
@@ -607,6 +607,31 @@ def assertWarnsRegexp(self, expected_regexp=""):
+        """
+        An extended version of the usual :meth:`unittest.TestCase.assertLogs`,
+        which also exercises the logger's message formatting.
+
```
```suggestion
        Also adds the 'msg_regex' kwarg :  
        If used, check that the result is a single message of the expected level, 
        and that it matches this regex.
```

pp-mo commented on lib/iris/tests/__init__.py:
```diff
@@ -607,6 +607,31 @@ def assertWarnsRegexp(self, expected_regexp=""):
+        # Check for any formatting errors by running all the formatters.
+        for record in watcher.records:
+            for handler in assertlogging_context.logger.handlers:
+                handler.format(record)
```
```suggestion
                handler.format(record)

        # Check message, if requested
        if msg_regex:
            self.asssertEqual(len(watcher.records), 1)
            rec = watcher.records[0]
            self.assertEqual(level, rec.levelname)
            self.assertRegex(rec.msg, msg_regex)
```

pp-mo reviewed (commented):
> > Don't want to undermine the work here, but I think the test changes are looking messy.

> 3cf6662 Does this help at all?

I really meant the way we formerly had extra code to 'exercise' the logging before capturing + checking it.
So that is all squared away now 👍 


> > Would you consider incorporating into this PR the above extension to the assertLogs method ??

> bfbed13 I can confirm from local testing that this addresses the issue you originally identified.

Brilliant, thanks ! 💐 


> > I do still have some concerns about the completeness of that aproach, though

> I get what you're saying, however this area is new to us and we have addressed all issues 
we've encountered so far. No problem IMO.

👍 I like your approach 😁 


One thing that does still jar here, for me, is the repetition of the 'check_log` method : we now have 4 copies (!)
I was very close to adding that (i.e. the regex check) into the new-version `assertLogs`, but I felt it was perhaps too much in one.
But what the hell, why not add an optional keyword like 'message_regex' + integrate the  two?
The only snag, IMO, is that it only applies when you're expecting a single logged message.  But naming/documentation can just allow for that.
See my suggestions (only partial changes : you will need to replace all the `check_log`s with `assertLogs` changes)

**How does that way look to you @trexfeathers ?**

trexfeathers commented:
>How does that way look to you @trexfeathers ?

Very nice idea @pp-mo! 5a997d333175ded17c2f9202f849a6cb2317e6c1 should take care of this - involved a fair amount of refactoring but the result is simpler, a little more akin to `AssertRaisesRegex()`, which is what I was originally after anyway.

pp-mo commented:
This all seems good to me now.
Thanks for sticking with it @trexfeathers
